### PR TITLE
fix: prevent watcher ingestion starvation

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary

- precompute live-file parent evidence once per poll so offset pruning is linear rather than O(tracked offsets × live files)
- cap unread JSONL suffixes with `BRAINLAYER_WATCH_MAX_FILE_BYTES` (100 MiB default), checkpoint skipped oversized suffixes immediately, and continue processing other files
- cover pruning complexity, cap configuration, checkpoint persistence, and continuation behavior with regression tests

## Root cause

The serial watcher could spend each poll in a quadratic `any(... is_relative_to(...))` prune scan across thousands of tracked files. Separately, one multi-gigabyte rollout suffix could be read without a bound and monopolize the watcher. Together these starved liveness and realtime inserts.

## Verification

- watcher suite: `88 passed`
- signed-safe baseline after rebase: `3672 passed, 9 skipped, 77 deselected, 1 xfailed`
- mandatory pre-push gate (with macOS fd ceiling raised to 4096): `3602 passed, 9 skipped, 61 deselected, 1 xfailed`; MCP registration `3 passed`; isolated eval/hook routing `40 passed`; Bun `1 passed`; FTS determinism passed
- Ruff on changed files: passed
- `git diff --check origin/main...HEAD`: passed

## Review note

Local CodeRabbit suggested retaining a missing tracked file when a live file exists beneath its parent. That is contrary to the existing prune contract: a live descendant is precisely the parent evidence that makes the missing candidate provably stale and removable. Existing and new tests cover direct-parent evidence, nested unavailable roots, and empty subtrees, so no semantic change was made for that suggestion.

## Operational scope

This PR changes source and tests only. It does not deploy or alter the live LaunchAgent, Homebrew Cellar, release tags, package artifacts, tap, or cask.

---

> [!NOTE]
> **Medium Risk**
> Changes core realtime ingestion offset semantics (skip/checkpoint of large pending suffixes and discarded-byte advancement); misconfiguration or edge cases around unconfirmed flushes could drop or delay data, though tests target those paths.
> 
> **Overview**
> Addresses watcher starvation from slow offset pruning and unbounded reads on huge rollout suffixes.
> 
> **Prune performance:** `OffsetRegistry.prune_missing_files` now builds a `live_parent_dirs` set once per cycle and uses membership checks instead of scanning every live file with `is_relative_to`, making pruning linear in tracked paths rather than quadratic.
> 
> **Oversized pending reads:** `BRAINLAYER_WATCH_MAX_FILE_BYTES` (default 100 MiB) limits how many unread bytes the watcher will pull from a file in one go. When pending data exceeds the cap, `_skip_oversized_file` checkpoints the registry to EOF (with immediate flush), logs a warning, and skips further reads for that file until a smaller append brings pending under the limit; `0` disables the cap.
> 
> **Poll loop behavior:** `poll_once` drains in-memory tailer buffers via `read_buffered_lines` before opening files (so buffered lines are not lost when an oversized append would otherwise short-circuit reads). Rewind side effects move to `_handle_rewind`. `_checkpoint_discarded_progress` advances confirmed offsets past bytes that were read but not indexable, without jumping past unconfirmed indexer batches (`has_buffered_source`).
> 
> Regression tests cover prune complexity, oversize skip/checkpoint, rewind-on-replace, and env validation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a4ed29d8debe52f1b2b10672ed5648b973ac7575. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Fix watcher ingestion starvation by skipping oversized files and draining buffered lines
> - Adds `_skip_oversized_file` to `JSONLWatcher`: when pending unread bytes exceed a configurable threshold (`BRAINLAYER_WATCH_MAX_FILE_BYTES`, default 100 MiB), the file is checkpointed to its end and skipped until it falls below the threshold.
> - Drains already-buffered complete lines before re-reading disk, preventing starvation when large files block the poll loop.
> - Extracts `JSONLTailer.read_buffered_lines` and `has_complete_buffered_line` so callers can consume buffered records without additional filesystem reads.
> - Adds `BatchIndexer.has_buffered_source` to prevent checkpointing past unconfirmed buffered entries during oversized-file skips.
> - Optimizes `OffsetRegistry.prune_missing_files` to use precomputed parent-directory sets instead of per-file linear scans.
> - Risk: files exceeding the byte threshold will have their unread content silently dropped; the checkpoint advances to end-of-file, skipping any unindexed records in that region.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized a4ed29d.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Added a configurable cap on pending unread data size for JSONL monitoring via environment setting.
  * When limits are exceeded, oversized files are skipped but their registry offsets advance and are checkpointed for safe later recovery.
* **Bug Fixes**
  * Improved pruning of missing offsets using more accurate live-parent evidence evaluation.
  * Enhanced rewind/checkpoint restore behavior for oversized scenarios with reliable rewind notifications.
* **Chores / Tests**
  * Expanded watcher tests to cover oversize skipping, checkpointing edge cases, and invalid/disabled limit configurations.

### CODE REVIEW

File: src/brainlayer/watcher.py [hunk 1]
```diff
@@ -1098,6 +1126,43 @@ def _ensure_tailer(self, filepath: str) -> JSONLTailer:
...
+            return False
+
+        try:
+            file_stat = os.stat(filepath)
+        except OSError:
+            return False
+
+        tailer = self._tailers.get(filepath)
+        offset = tailer.offset if tailer else self.registry.get(filepath)[0]
+        pending_bytes = max(file_stat.st_size - offset, 0)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/watcher.py:1218]: Fixed in bf7e85e0. The cap now compares the persisted registry inode with the current stat inode and treats a mismatch as offset 0 before computing pending bytes. I used the registry inode because JSONLTailer.get_inode() stats the current pathname and cannot represent the old tailer inode after replacement. Added a RED→GREEN atomic-replacement regression test; watcher suite is 89 passed and the full mandatory pre-push gate passed.

File: src/brainlayer/watcher.py [hunk 2]
```diff
@@ -1098,6 +1126,43 @@ def _ensure_tailer(self, filepath: str) -> JSONLTailer:
...
+                "pending_bytes=%d max_file_bytes=%d offset=%d size=%d",
+                filepath,
+                pending_bytes,
+                self.max_file_bytes,
+                offset,
+                file_stat.st_size,
+            )
+        self._oversized_files.add(filepath)
+        return True
+
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/watcher.py:1266]: Fixed in 704f598f. Each poll now intersects `_oversized_files` with currently discovered, non-denylisted paths before file processing. Added a RED→GREEN test covering both disappearance and a path becoming denylisted. Watcher suite is 90 passed; full mandatory pre-push gate passed.

File: src/brainlayer/watcher.py [hunk 3]
```diff
@@ -1098,6 +1126,45 @@ def _ensure_tailer(self, filepath: str) -> JSONLTailer:
...
+        offset = tailer.offset if tailer else registry_offset
+        if registry_inode != 0 and registry_inode != file_stat.st_ino:
+            offset = 0
+        pending_bytes = max(file_stat.st_size - offset, 0)
+        if pending_bytes <= self.max_file_bytes:
+            self._oversized_files.discard(filepath)
+            return False
+
+        self._tailers.pop(filepath, None)
+        self.registry.set(filepath, file_stat.st_size, file_stat.st_ino)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/watcher.py:1249]: Fixed in 794d5ce6. Before an oversized EOF checkpoint, the watcher now force-flushes buffered entries from that source and defers checkpointing while any remain retained/unconfirmed. Added a RED→GREEN flush-failure test proving the registry stays at its prior offset and the buffered item remains available. Watcher suite is 92 passed; full mandatory pre-push gate passed.

File: src/brainlayer/watcher.py [hunk 4]
```diff
@@ -1098,6 +1126,45 @@ def _ensure_tailer(self, filepath: str) -> JSONLTailer:
...
+            file_stat = os.stat(filepath)
+        except OSError:
+            return False
+
+        tailer = self._tailers.get(filepath)
+        registry_offset, registry_inode = self.registry.get(filepath)
+        offset = tailer.offset if tailer else registry_offset
+        if registry_inode != 0 and registry_inode != file_stat.st_ino:
+            offset = 0
+        pending_bytes = max(file_stat.st_size - offset, 0)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/watcher.py]: Fixed in 794d5ce6. `_skip_oversized_file` now treats `size < tailer offset` as a byte-zero rewind, and calls `mark_rewind` before persisting the smaller EOF so the registry generation safely permits the lower same-inode checkpoint. Added a RED→GREEN same-inode truncation test.

File: src/brainlayer/watcher.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/watcher.py:1249]: Addressed in 794d5ce6 by the same pending-entry guard: if this source has buffered entries, the watcher force-flushes them and rechecks the buffer; while any remain retained after a full or partial flush failure, it returns without advancing the registry. The RED→GREEN regression asserts the prior registry offset and retained item are preserved. Successful production flushes return confirmed watermarks through `_advance_confirmed_offsets` before the EOF checkpoint.

File: src/brainlayer/watcher.py [hunk 5]
```diff
@@ -1098,6 +1131,59 @@ def _ensure_tailer(self, filepath: str) -> JSONLTailer:
...
+                        "Oversized JSONL checkpoint deferred for unconfirmed entries: %s",
+                        filepath,
+                    )
+                self._oversized_files.add(filepath)
+                return True
+
+        self._tailers.pop(filepath, None)
+        if file_rewound:
+            self.registry.mark_rewind(filepath, file_stat.st_ino)
+        self.registry.set(filepath, file_stat.st_size, file_stat.st_ino)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/watcher.py:1249]: Fixed in 3c5f70a5. Rewind handling is now centralized so both ordinary reads and oversized checkpoints execute the same registry generation, telemetry, and `on_rewind` archival callback before the EOF checkpoint. The same-inode oversized rewind RED test now asserts the callback receives the exact old/new offsets. Watcher suite is 92 passed; full mandatory pre-push gate passed.

File: src/brainlayer/watcher.py [hunk 6]
```diff
@@ -1098,6 +1131,103 @@ def _ensure_tailer(self, filepath: str) -> JSONLTailer:
...
+            return False
+
+        try:
+            file_stat = os.stat(filepath)
+        except OSError:
+            return False
+
+        tailer = self._tailers.get(filepath)
+        registry_offset, registry_inode = self.registry.get(filepath)
+        offset = tailer.offset if tailer else registry_offset
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/watcher.py]: Fixed in 2001b074. The cap is now anchored to the durable registry watermark rather than the volatile tailer offset. If a same-inode tailer remains ahead after a forced flush, EOF checkpointing is deferred even when the batch buffer is empty. Added a RED→GREEN two-line partial-confirmation test proving the lower confirmed offset is preserved after an oversized append. Watcher suite is 93 passed; full mandatory pre-push gate passed.

File: src/brainlayer/watcher.py [hunk 7]
```diff
@@ -1098,6 +1131,110 @@ def _ensure_tailer(self, filepath: str) -> JSONLTailer:
...
+            return False
+
+        try:
+            file_stat = os.stat(filepath)
+        except OSError:
+            return False
+
+        tailer = self._tailers.get(filepath)
+        registry_offset, registry_inode = self.registry.get(filepath)
+        tailer_offset = tailer.offset if tailer else registry_offset
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/watcher.py:1213]: Fixed in 7890d188. The watcher now separates parsing already-buffered complete records from reading additional file bytes. When a tailer has buffered records, poll_once drains them under max_lines_per_file without calling f.read(), while preserving inode-replacement and rewind handling. RED reproduced the EOF checkpoint/data loss; GREEN: 94 watcher tests and the full pre-push gate (3608 passed, 9 skipped, 1 xfailed).

File: src/brainlayer/watcher.py [hunk 8]
```diff
@@ -1098,6 +1139,110 @@ def _ensure_tailer(self, filepath: str) -> JSONLTailer:
...
+        pending_bytes = max(file_stat.st_size - offset, 0)
+        if pending_bytes <= self.max_file_bytes:
+            self._oversized_files.discard(filepath)
+            return False
+
+        if tailer is not None and not inode_changed and not file_rewound and tailer_offset > registry_offset:
+            if self.indexer.has_buffered_source(filepath):
+                self.indexer.flush()
+            confirmed_offset, confirmed_inode = self.registry.get(filepath)
+            if confirmed_inode != file_stat.st_ino or confirmed_offset < tailer_offset:
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/watcher.py:1227]: Fixed in 9be818ad. The watcher now checkpoints intentionally discarded progress only when the registry has already confirmed every preceding indexable record and no source entry remains buffered. It therefore prevents the cap from deferring forever after dropped-only records without ever jumping past mixed-stream unconfirmed work. RED/GREEN regressions cover both cases; watcher suite 96 passed, full pre-push gate 3610 passed.
